### ISSUE TITLE
RLHF

### ISSUE THREAD

**Issue by 0x4007** [ubiquity-os-marketplace/command-ask#18]
> The acronym is RLHF, which stands for Reinforcement Learning from Human Feedback. This method involves humans guiding the behavior and outputs of the model by providing feedback on its responses, helping it learn desired responses over time.

We should support three inputs for the model to learn from:
1. Thumbs up
2. Thumbs down
3. Edits

Edits could be useful to compare what the output should be. That way we have a chance to correct some specific mistakes. In the below example it incorrectly writes the `.ubiquity-os.config.yml` name and this could be a good opportunity to teach it. 

https://github.com/ubiquity-os/ubiquity-os-kernel/issues/111#issuecomment-2439978163

**Comment by 0x4007** [comment]
@sshivaditya2019 Rfc

**Comment by Keyrxng** [comment]
This feels like the wrong repo to be opening this task in.

1. We'd need to store the AI response and using the DB comes with overheads.
2. We should use `structuredMetadata` like I've implemented in #14 for posting ground truths.
3. When the time comes we write a script to parse all of our LLM response comments and we extract the comment metadata which contains the reactions to the comment and edits (maybe need GQL to fetch these)
4. Sanitize and prep the dataset and then train our own model.

**Comment by 0x4007** [comment]
Simpler to just compare the earliest and latest comment revision. Don't think we need to complicate this and store in database or generate any embeddings. Just add diff to context and say 

```
For this example output 
(Before)
Make sure to adjust the response to accommodate 
(After diff) 
```

Something like that.

**Comment by 0x4007** [comment]
Simpler to just compare the earliest and latest comment revision. Don't think we need to complicate this and store in database or generate any embeddings. Just add diff to context and say 

```
For this example output 
(Before)
Make sure to adjust the response to accommodate 
(After diff) 
```

Something like that.

**Comment by Keyrxng** [comment]
I may be out of my depth but we need to move away from thinking in terms of prompt injection when considering model training. We need to fine-tune a model with specific datasets to embed it into it's foundational knowledge separate from our prompts.

This training will also be partner specific so we should build with this in mind so we can white label it as a service for partner' chatbot.

**Comment by 0x4007** [comment]
First step is always prompting.

Later comes RAG/true fine tuning. 

Technically this may not qualify as true RLHF

**Comment by Keyrxng** [comment]
> we could keep a track of the reactions and edits, we keep track of edit diffs as well

GitHub does this for us already so we should not use our own DB to handle that aspect

> We can monitor both positive and negative examples within a repository and incorporate them into the prompt at query time.

This would imply that for all correction and edits that are made repository wide they are going to be pulled from a DB (I assumed you meant keeping track via DB storage) and they'd all be fed into the context window?

---

1. No Supabase it's not required we can use GitHub alone if we wanted to.
2. We use `structuredMetadata` to create a trackable header per LLM response comment, have it contain key details: `comment_id` it replied to for easy indexing, models used, tokens, groundTruths, application (e.g chatbot or codeReview)
3. We either A) use Search API for the comment header in that repo B) list all comments and use the parser.
4. The `comment` obj contains the reactions on the comment (perhaps restrict one to regenerate the response with?)


I don't think injecting all of the edits etc into the systemMessage is the way to go with it exactly if we consider our issues with context window so far. Perhaps we build another LLM call like for groundTruths and have it summarize or succinctly embody all of the various `RLHF` techinques we intend for this feature?

Our system message is already MASSIVE and littered with context from all sorts of tasks, I don't think injecting 10s of diffs and full body llm responses and the original context window it was fed is a great idea.

**Comment by Keyrxng** [comment]
Appreciate the response that clarified things for me, I'm looking forward to seeing it implemented.

**Comment by 0x4007** [comment]
How are you dealing with the numbers? Assuming using typescript because if it's being handled directly by the LLM it wont be great.
